### ISSUE TITLE
Dynamic comma

### ISSUE THREAD

**Issue by ksv-success** [cocus/vc288-alt-fw#7]
если я правильно понял с запятой нужно работать здесь? 

void set_segments(uint8_t data)
{
    PA_ODR_ODR2 = data & 1; data >>= 1;     // G
    PC_ODR_ODR4 = data & 1; data >>= 1;     // F
    PC_ODR_ODR7 = data & 1; data >>= 1;     // E
    PC_ODR_ODR6 = data & 1; data >>= 1;     // D
    PC_ODR_ODR3 = data & 1; data >>= 1;     // C
    PA_ODR_ODR3 = data & 1; data >>= 1;     // B
#ifndef SWIM_DEBUG_ENABLED
    PD_ODR_ODR1 = data & 1; data >>= 1;     // A (same pin of SWIM)
#else
    data >>= 1;
#endif
     PC_ODR_ODR5 = data & 1; data >>= 1;     // Dp       
}

void select_digit(uint8_t digit)
{
    PD_ODR_ODR4 = digit != 0;
    PD_ODR_ODR6 = digit != 1;
    PD_ODR_ODR5 = digit != 2;

    PB_ODR_ODR5 = digit != 3;
    PB_ODR_ODR4 = digit != 4;
    PA_ODR_ODR1 = digit != 5;
}

**Comment by cocus** [comment]
No, the 8th bit of the argument "data" of that function will actually decide if the decimal place is lit or not. There are no access to that since it's on the TO-DO list.
Maybe in the upcoming future I can add that.

**Comment by ksv-success** [comment]
Хорошо спасибо буду ждать будущее

**Comment by cocus** [comment]
Please have a look at this branch.

https://github.com/cocus/vc288-alt-fw/tree/cocus/calibration-and-moving-decimal-point-test

Note that if you changed any of the "calibrations" you should review these in this code.
In order to calibrate the offset (i.e. reading non zero values when nothing is applied), you should disconnect any inputs, and tie the Voltage input to ground. Short the J2 and then power on the device. After a moment, release the short and after a moment it should do it.

Dynamic comma is also applied, but not tried. I don't have my meter at hand to try it out.

**Comment by cocus** [comment]
Hello, did you manage to try this? I'm interested on some feedback!

**Comment by ksv-success** [comment]
Спасибо! извините что так долго отвечал, только увидел сообщение на почте, попробую сегодня или завтра и напишу что у меня вышло )

**Comment by ksv-success** [comment]
все так же нет запятой к сожалению...

**Comment by ksv-success** [comment]
> Пожалуйста, посмотрите на эту ветку.
> 
> https://github.com/cocus/vc288-alt-fw/tree/cocus/calibration-and-moving-decimal-point-test
> 
> Обратите внимание, что если вы изменили какую-либо из «калибровок», вам следует просмотреть их в этом коде.
> Чтобы откалибровать смещение (т. Е. Считать ненулевые значения, когда ничего не применяется), вы должны отсоединить все входы и привязать вход напряжения к земле. Замкните J2 и включите устройство. Через мгновение отпустите короткое и через мгновение оно должно это сделать.
> 
> Динамическая запятая также применяется, но не пробуется. У меня нет под рукой моего счетчика, чтобы опробовать его.

сделал так как вы сказали но ничего не изменилось все так же как и было

**Comment by cocus** [comment]
even if you supply some voltage to the input? Have you calibrated it before starting?

**Comment by ai-big-boss** [comment]
я калибровал по иструкции ничего не изменилось код не менял все показывает как раньше 077 и 097 
напряжение не измерял сейчас попробую измерить

**Comment by ai-big-boss** [comment]
извините думал начальный код нужно установить я идиот ))) теперь понял но пока не понял что скодом точки делать эти все файлы нужно заменить?

**Comment by ai-big-boss** [comment]
я устанавливал опять же этот код https://github.com/cocus/vc288-alt-fw а их нужно обеденить если правильно понял в один с этим https://github.com/cocus/vc288-alt-fw/tree/cocus/calibration-and-moving-decimal-point-test

**Comment by ksv-success** [comment]
https://github.com/cocus/vc288-alt-fw/tree/cocus/calibration-and-moving-decimal-point-test
совсем ничего не горит даже после калибровки

**Comment by ksv-success** [comment]
вообщем никаких показаний не показывает даже не горит не один сегмент ни до калибровки ни после возможно потому что у меня STM8S103F3P ничего не работает хотя первый код работал вполне нормально без каких либо изменений

**Comment by ksv-success** [comment]
нет индикации совсем

**Comment by ksv-success** [comment]
изменил значение 0 на 1 
gpio.c 
"uint8_t read_programming_pin(void)
{
    return PD_IDR_IDR6 == 1; 
}"
и появилась индикация амперметра показывает 97.0 A... 
Вольтметр не показывает пока не подключишь измерение напряжения

**Comment by ai-big-boss** [comment]
![IMG_20200123_181848](https://user-images.githubusercontent.com/59658638/72997287-300ee180-3e0d-11ea-9930-6f7fae3f112a.jpg)

**Comment by ksv-success** [comment]
![1579793486361499439349](https://user-images.githubusercontent.com/59658638/72998381-c8f22c80-3e0e-11ea-9cf2-30de5f49f2b0.jpg)

**Comment by ksv-success** [comment]
Заработало осталось только откалибровать

**Comment by ksv-success** [comment]
не калибруется думаю нужно поставить резисторы правильного номинала

**Comment by ksv-success** [comment]
калибровка не работает(

**Comment by ai-big-boss** [comment]
/* Raw value from ADC */
    value = get_adc_volts();
    /* Offset */
    value -= settings->adc_volts.offset;
    /* Scale */
    value *= settings->adc_volts.scaling;
    /* Clamp to zero */
    if (value < 0.0)
    {
        set_display_from_int(0, 0, 1);
    }
    else if (value >= 100.0)
    {
        /* 100. - 999. */
        set_display_from_int(value, 0, 4);
    }
    else if (value >= 10.0)
    {
        /* 10.0 - 99.9 */
        set_display_from_int(value*10, 0, 2);
    }
    else if (value >= 1.0)
    {
        /* 1.00 - 9.99 */
        set_display_from_int(value*100, 0, 1);
    }
}

**Comment by cocus** [comment]
If you change the 1 to 0 as you've mentioned, you're bypassing the calibration routine.
You need to short that pin briefly and wait a little bit. After that it should calibrate.

**Comment by ksv-success** [comment]
else if (value >= 0.001)
    {
        /* 0.001 - 0.999 */
        set_display_from_int(value*1000, 3, 1);
    }


возможно ли добавить такой код чтобы показывал значения до 1 Вольта

**Comment by ksv-success** [comment]
замыкаю j2 и контакт измерений на землю и подаю питание  загорается индикатор А центральный вольтметра после больше 30 секунд отпускаю и ничего

**Comment by ksv-success** [comment]
![IMG_20200124_195101_HHT](https://user-images.githubusercontent.com/59658638/73087073-ef819780-3ee2-11ea-8a38-f1caab1eb653.jpg)

**Comment by ksv-success** [comment]
Я уже и резистор 270кОм заменил на 300 кОм на амперметре так как на вольтметре тоже 300 кОм возможно нужно точное значение измерить мультиметром и добавить в программу ?

**Comment by ksv-success** [comment]
спалил программатор при попытке замерить напряжения 12 вольт после калибровки теперь у программатора 3 вольта закорочено на землю возможно что программа сработала неправильно?  но как через резистор 300 кОм могло поступить напряжение на линию 3 вольта странно

**Comment by ksv-success** [comment]
ну вообщем спасибо за вашу работу не бросайте я все же спустя долгое время смог восстановить свой испорченный вольтметр благодаря вам и вашей программе после ремонта программатора или покупки нового буду опять пробовать вашу программу надеюсь STM8 жива )) может быть изначально с программатором проблема была криво устанавливал программу вообщем до встречи )

**Comment by cocus** [comment]
Hi, I've updated the branch. Please download it again. Note that on the first time you use this, you need to comment out line 170 out of main.c; and a calibration will take place when you start the meter. After that, please un-comment it. A proper calibration trigger will be added later. Dynamic comma should work "ok" now, and the amps should display something useful as well. Although the gain value for the amps is not right.

**Comment by ksv-success** [comment]
Привет! хорошо обновлю сейчас только недавно вроде бы восстановил свой программатор вышедший из строя после неудачного замера напряжения сгорел микроконтроллер stm32 заменил прошил буду пробывать заказа новый с Китая теперь будет 2 программатора ))

**Comment by ksv-success** [comment]
проэкт не компилируется ошибки выбивают но я пробывал старую версию вообщем и STM8 умерла буду ждать с китая STM8 закажу с излишком штук 10 мало ли ))) как получу и перепаяю установлю вашу программу и напишу обязательно

**Comment by ksv-success** [comment]
![image](https://user-images.githubusercontent.com/59658638/76868248-a8c37480-6877-11ea-9040-7398278b12fb.png)
https://github.com/cocus/vc288-alt-fw/tree/cocus/calibration-and-moving-decimal-point-test

**Comment by cocus** [comment]
Good catch! I forgot to push that file! Please clone again, I've pushed that file.

**Comment by ksv-success** [comment]
![image](https://user-images.githubusercontent.com/59658638/76950546-3c01b600-691b-11ea-8c07-eaf3753230dc.png)
или мой программатор неисправен или stm8 буду разбираться в чем проблема спасибо вам защищайтесь от короновируса ) вы нужный человек!

**Comment by cocus** [comment]
I'm not sure why that happens. do you mind cleaning all and then trying to just build and if that worsk, then debug?

**Comment by ksv-success** [comment]
Спасибо за ожидание получил программатор новый ! прошил новый stm8f003f3 все супер запятая есть все работает ! показывает 2.7 вольт при замере 3.3 ну это не критично. Спасибо огромное за ваш труд и поддержку !

**Comment by ksv-success** [comment]
![IMG_20200413_165103](https://user-images.githubusercontent.com/59658638/79125823-7853f500-7da7-11ea-8d73-cab386905f0f.jpg)
Странные показания посторонние

**Comment by ksv-success** [comment]
![IMG_20200413_165100](https://user-images.githubusercontent.com/59658638/79125914-99b4e100-7da7-11ea-9b41-d6d6a68b7197.jpg)
Как обнулить амперметр ? Я пробывал изменять настройки которые вы говорили в прошлой версии но мало что изменяется

**Comment by ksv-success** [comment]
Амперы показывает на оборот не +10 а в -10 то есть должен показывать например 10 ампер а он показывает -10

**Comment by ksv-success** [comment]
Значения идут в минус

**Comment by ksv-success** [comment]
![IMG_20200414_110611](https://user-images.githubusercontent.com/59658638/79201052-1e533e00-7e40-11ea-93e8-aec7a0148e54.jpg)
![IMG_20200414_110603](https://user-images.githubusercontent.com/59658638/79201063-214e2e80-7e40-11ea-8427-de3e948c4ed9.jpg)
Изменил значение семплов на 2 вроде бы показывает почти 3.3 вольта но обманывает на 0.10 вольт не критично но все же что сделать чтобы показывало верно от 0.01 вольта

**Comment by ksv-success** [comment]
What do I need to do to calibrate?

**Comment by ksv-success** [comment]
only the change #define ADC_AVERAGE_SAMPLES_BASE_2 5

**Comment by ksv-success** [comment]
other settings do not change readings, all readings remain unchanged

**Comment by ksv-success** [comment]
what needs to be done so that the readings after the decimal point smoothly change, for example, from 0.01 to 0.99 (0.02, 0.03, 0.04 and so on)

**Comment by ksv-success** [comment]
Is calibration possible once with firmware?

**Comment by ksv-success** [comment]
Sorry I will wait for an answer dear friend

**Comment by ksv-success** [comment]
![image](https://user-images.githubusercontent.com/59658638/79308455-4b1b5a00-7f01-11ea-928d-f9c7fa05db8a.png)

**Comment by ksv-success** [comment]
когда замыкаю JP2 минуту удерживаю показания не изменяются буду пробывать подбирать резисторы R2 и R3 но буду рад если ответите спасибо вам!

**Comment by ksv-success** [comment]
> Я не знаю, почему это происходит. Вы возражаете очистки всех, а затем пытается просто построить, и если это worsk, то отладить?

my ST-Link was faulty I ordered a new one in China, that's just got it and immediately started to flash, the first one for some reason burned out, I restored it STM32 firmware but does not want to flash STM8 :D , there is apparently something else failed, so this message was I now have everything in order the new programmer works properly.

**Comment by cocus** [comment]
So, the calibration doesn't start by itself. you need to remove the input voltage (i.e. the one you'd measure) and then flash it with main.c:173 commented. This will run the calibration ALWAYS. After it boots for the first time, revert this change (i.e. un-comment the same line) and then it should work.
I haven't figured a good value for the amps gain, which means that you'll read COUNTS from the ADC rather than amps. if you figure out the ratio (i.e. 1A on the input means for instance 300 counts, then the ratio is 1/300, which you could change on GAIN_FOR_AMPS and the repeat these steps).

**Comment by ksv-success** [comment]
а почему показания не изменяются когда я меняю значения в 
#define AMPS_R3_RV2                     200000.0

#define VOLTS_R2_RV1                    301500.0

я их изменил , но это ничего не изменило

**Comment by ksv-success** [comment]
я заметил погрешность в 1-2 вольт  при изменении напряжения от 1 до 20 вольт

**Comment by cocus** [comment]
Because those are the reference values that will be stored on the calibration. so if you change those defines, you need to "re-calibrate" your device. That small error you saw on the voltage on that range I think its due to the non linearity that this ADC shows (which I didn't do anything to fix it for now).

**Comment by ksv-success** [comment]
понял буду пробовать тестировать но уже делал замену резисторов)) на вольтметре вроде бы удачно а вот амперметр бог знает чет непонятное на нем ))) что то показывает какие то числа когда ничего не подключено )))

амперметр показывает постоянно разные числа 200, 185, 197  очень быстро иногда удается разглядеть и при подключении нагрузки , измеряет в обратном направлении то есть если показывает 200 становится 180 то есть показания идут как бы в минус

**Comment by cocus** [comment]
Okay, the amp wasn't tried that much, but it could have some issues. Next time I have time to work on it will try to make the amp meter to work properly. Probably the formula for the gain is wrong. I'll have to debug it more closely. But the volt meter should work "kind of ok" for now.

**Comment by ksv-success** [comment]
если правильно понял после того как я уберу комментирование //while(!read_programming_pin()); 
 будет работать калибровка замыкания JP2

**Comment by cocus** [comment]
ehhh technically it should work from the beginning but I catched an issue with that pin. it seems like the LEDs connect from the SWIM line to that pin, and if you ground it, you'll pull the SWIM line low, which would break your debug session. I think it might work if you're not debugging it, but I think it won't work regardless. That's an issue with the design of the board. I will see what other methods I can come up with for triggering a calibration.

**Comment by ksv-success** [comment]
Хорошо буду ждать дополнений ! и тестировать то что вы сказали спасибо вам огромное!

**Comment by ksv-success** [comment]
настроил вольтметр подбором резисторов показывает правильно ! только после запятой ошибается и считает не от 1.00 до 1.99 а хаотично 1.23 , 1.11, 1.37 как то так хотя мультиметр показывает четко где можно настроить это исправить значения после запятой?

**Comment by ksv-success** [comment]
Где можно в программе изменить значения RV1 и R2 потому что у меня другое значение и найти точно такой же резистор с маркировкой 274 не могу поэтому поставил с маркировкой 304 
подобрал 200k и 75k но я так понял нужно ровно 291k... а возможно что бы калибровка шла от опорного напряжения ? там же идет на питание 3.3 вольта ровно или калибровка так и работает может я тупой жаль плохо понимаю английский ))) без переводчика да и с переводчиком тоже плохо)))

**Comment by ksv-success** [comment]
теперь я понял отредактировал main.c показывает 3.3 вольта ура! )) спасибо вам ! переведу ваши комментарии в коде на русский выложу вам может добавите в ваш код что бы другие понимали о чем речь кто плохо понимает английский как я ))))
![image](https://user-images.githubusercontent.com/59658638/79436633-cea87a00-7fd9-11ea-869d-29c2d069c787.png)

**Comment by ksv-success** [comment]
@cocus я получил сегодня новый вольтамперметр с Китая на STM8  :D

**Comment by ksv-success** [comment]
![IMG_20200417_130145](https://user-images.githubusercontent.com/59658638/79693023-09780f80-8271-11ea-8a2f-261247f45538.jpg) 
Вот этот новый китайский VC 288 на STM8 xD значит Naviton аналог Китайский 😕

**Comment by cocus** [comment]
Hey good, I don't speak russian myself and seems like english is a good "global" language for all projects, I hope you can translate those. Cool to see a proper STM8 meter.
The values for the resistors and such are on the config_8s003.h file, with comments on what each parameter means. You can skew the values by tweaking those potentitometers, but I picked a value (on the code) where they're on their middle point.

**Comment by ksv-success** [comment]
я взял новый китайский VC288 на котором была установлена STM8 и прошил его вашей прошивкой xD. Скажите как настроить его что бы он вел расчет правильно где настроить это? 
Он отображает вольтаж 1.00 и перескакивает до 1.11 , 1.22 и так далее., не плавно отображает значения где это можно настроить и какие значения нужно изменять? изначально он показывает нормально и более мение точно но он не показывал значения в 0.01 вольта минимум 0.1 вольт а ваш прошивка лучше но исправить бы амперметр и вольтметр чуть и будет лучше чем оригинал !

**Comment by ksv-success** [comment]
![IMG_20200424_091426](https://user-images.githubusercontent.com/59658638/80182892-89ef9500-8610-11ea-91cc-3a40d6a9c81b.jpg)
![IMG_20200424_091132](https://user-images.githubusercontent.com/59658638/80182902-8d831c00-8610-11ea-9b08-50ff9fc81aa4.jpg)
Ничего не меняется в чем причина может быть не программист к сожалению может что нибудь не так с этими настройками ----> config_8s003.h

**Comment by ksv-success** [comment]
странно почему эта формула не работает ? 
#define GAIN_FOR_VOLTAGE  (((VOLTS_R2_RV1 + VOLTS_R1) / VOLTS_R1) * ADC_SCALE)
я ее изменяю до такого состояния   
#define GAIN_FOR_VOLTAGE  ADC_SCALE
и ничего не меняется, показания остаются прежними это странно

**Comment by ksv-success** [comment]
Hello! how can I configure the ADC so that it shows 3.30 volts and not 3.17 - 3.29
